3-UsingAPIs/Week1/homework/ex3-rollDice.js

```diff
@@ -9,47 +9,38 @@
   explanation? Add your answer as a comment to be bottom of the file.
 ------------------------------------------------------------------------------*/
 
-// TODO Remove callback and return a promise
-function rollDice(callback) {
-  // Compute a random number of rolls (3-10) that the dice MUST complete
-  const randomRollsToDo = Math.floor(Math.random() * 8) + 3;
-  console.log(`Dice scheduled for ${randomRollsToDo} rolls...`);
-
-  const rollOnce = (roll) => {
-    // Compute a random dice value for the current roll
-    const value = Math.floor(Math.random() * 6) + 1;
-    console.log(`Dice value is now: ${value}`);
-
-    // Use callback to notify that the dice rolled off the table after 6 rolls
-    if (roll > 6) {
-      // TODO replace "error" callback
-      callback(new Error('Oops... Dice rolled off the table.'));
-    }
-
-    // Use callback to communicate the final dice value once finished rolling
-    if (roll === randomRollsToDo) {
-      // TODO replace "success" callback
-      callback(null, value);
-    }
-
-    // Schedule the next roll todo until no more rolls to do
-    if (roll < randomRollsToDo) {
-      setTimeout(() => rollOnce(roll + 1), 500);
-    }
-  };
-
-  // Start the initial roll
-  rollOnce(1);
+function rollDice() {
+  const rollPromise = new Promise((resolve, reject) => {
+    // Compute a random number of rolls (3-10) that the dice MUST complete
+    const randomRollsToDo = Math.floor(Math.random() * 8) + 3;
+
+    console.log(`Dice scheduled for ${randomRollsToDo} rolls...`);
+    const rollOnce = (roll) => {
+      // Compute a random dice value for the current roll
+      const value = Math.floor(Math.random() * 6) + 1;
+      console.log(`Dice value is now: ${value}`);
+
+      if (roll > 6) {
+        reject(new Error('Oops... Dice rolled off the table.'));
+      }
+      if (roll === randomRollsToDo) {
+        resolve(`Success! Dice settled on ${value}.`);
+      }
+      if (roll < randomRollsToDo) {
+        setTimeout(() => rollOnce(roll + 1), 500);
+      }
+    };
+    // Start the initial roll
+    rollOnce(1);
+  });
+  return rollPromise;
 }
 
-// TODO Refactor to use promise
-rollDice((error, value) => {
-  if (error !== null) {
-    console.log(error.message);
-  } else {
-    console.log(`Success! Dice settled on ${value}.`);
-  }
-});
+rollDice()
+  .then((resolveValue) => console.log(resolveValue))
+  .catch((rejectValue) => console.log(rejectValue.message));
 
+// It looks like the promise gives the value (reject or resolve) immediately after
+// the first case (rejection or resolving) happen and doesn't change this value anymore.
```

Comment:
Yes, that is correct. Once a promise is 'settled' (i.e., resolved or rejected) its outcome can no longer be changed. Promises are deliberately designed to work that way.